src/modules/openjtalk.c

```diff
@@ -0,0 +1,237 @@
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <unistd.h>
+#include <string.h>
+#include <assert.h>
+#include <linux/limits.h>
+
+#include "module_main.h"
+#include "module_utils.h"
+
+#define MODULE_NAME "open_jtalk"
+#define MODULE_VERSION "0.1"
+
+DECLARE_DEBUG();
+
+MOD_OPTION_1_STR(OpenjtalkDictionaryDirectory);
+MOD_OPTION_1_STR(OpenjtalkVoice);
+
+int module_init(char **msg)
+{
+	fprintf(stderr, "initializing\n");
+
+	*msg = strdup("ok!");
+
+	return 0;
+}
+
+int module_load(void)
+{
+	INIT_SETTINGS_TABLES();
+	REGISTER_DEBUG();
+
+	MOD_OPTION_1_STR_REG(OpenjtalkDictionaryDirectory,
+			     "/var/lib/mecab/dic/open-jtalk");
+	MOD_OPTION_1_STR_REG(OpenjtalkVoice,
+			     "/usr/share/hts-voice/nitech-jp-atr503-m001/nitech_jp_atr503_m001.htsvoice");
+
+	DBG("OpenjtalkDictionaryDirectory: %s", OpenjtalkDictionaryDirectory);
+	DBG("OpenjtalkVoice: %s", OpenjtalkVoice);
+
+	module_audio_set_server();
+
+	return 0;
+}
+
+SPDVoice **module_list_voices(void)
+{
+	/* TODO: Allow users to choose htsvoice */
+	SPDVoice **ret = malloc(2 * sizeof(*ret));
+
+	ret[0] = malloc(sizeof(*(ret[0])));
+	ret[0]->name = strdup("Default");
+	ret[0]->language = strdup("ja");
+	ret[0]->variant = NULL;
+
+	ret[1] = NULL;
+
+	return ret;
+}
+
+void module_speak_sync(const char *data, size_t bytes, SPDMessageType msgtype)
+{
+	module_speak_ok();
+
+	DBG("speaking '%s'", data);
+
+	module_report_event_begin();
+
+	/* Strip SSML (Open JTalk does not support it.) */
+	char *plain_data = module_strip_ssml(data);
+
+	char template[] = "/tmp/speechd-openjtalk-XXXXXX";
+	int tmp_fd = mkstemp(template);
+	if (tmp_fd == -1) {
+		DBG("temporary .wav file creation failed");
+		goto FINISH;
+	}
+
+	/* do not need fd */
+	close(tmp_fd);
+
+	/* construct command line */
+	char cmd[PATH_MAX * 4];
+	snprintf(cmd, sizeof(cmd),
+		 "open_jtalk -x %s -m %s -ow %s", OpenjtalkDictionaryDirectory,
+		 OpenjtalkVoice, template);
+
+	FILE *oj_fp = popen(cmd, "w");
+	if (oj_fp == NULL) {
+		DBG("failed to execute open_jtalk");
+		goto FINISH;
+	}
+
+	fprintf(oj_fp, "%s", plain_data);
+
+	/* wait for finish */
+	int status = pclose(oj_fp);
+	if (status != 0) {
+		DBG("open_jtalk exited with non-zero code");
+		goto FINISH;
+	}
+
+	/* play the output wav */
+	DBG("output to %s", template);
+
+	AudioTrack track = {
+		.bits = 0,
+		.num_channels = 0,
+		.sample_rate = 0,
+		.num_samples = 0,
+		.samples = NULL
+	};
+	AudioFormat format = SPD_AUDIO_LE;
+
+	FILE *audio_fp = fopen(template, "rb");
+	if (audio_fp == NULL) {
+		DBG("failed to open wav file");
+		goto FINISH;
+	}
+	DBG("opened wav file");
+
+	fseek(audio_fp, 34, SEEK_SET);
```

Comment:
Please make these 34, 22, etc. macros defined at the top of the file, so they get meaningful naming, and their definitions are in one place.